File: ios/PitchDetectorModule.swift
//
//  PitchDetectorModule.swift
//  PitchDetector
//
//  Created by Fábio Pereira on 03/04/23.
//  Copyright © 2023 PitchDetector. All rights reserved.
//

import Foundation

@objc(PitchDetectorModule)
class PitchDetectorModule: NSObject {
  private let processor = BaseProcessor()
  private var recording = false

  @objc(start:reject:)
    func start(resolve:@escaping RCTPromiseResolveBlock,reject:@escaping RCTPromiseRejectBlock) -> Void {
    if (!self.recording) {
        let promise = PromiseUtils(resolve, reject)
        do {
            processor.start()
            self.recording = true
            promise.resolve(nil)
        } catch {
            self.recording = false
            promise.reject(error as NSError)
        }
    }
  }
    
  @objc(stop:reject:)
    func stop(resolve:@escaping RCTPromiseResolveBlock,reject:@escaping RCTPromiseRejectBlock) -> Void {
    if (self.recording) {
        let promise = PromiseUtils(resolve, reject)
        do {
            processor.stop()
            self.recording = false
            promise.resolve(nil)
        } catch {
            self.recording = false
            promise.reject(error as NSError)
        }
    }
  }

  @objc(isRecording:reject:)
    func isRecording(resolve:@escaping RCTPromiseResolveBlock,reject:@escaping RCTPromiseRejectBlock) -> Void {
    let promise = PromiseUtils(resolve, reject)
    do {
        promise.resolve(self.recording)
    } catch {
        promise.reject(error as NSError)
    }
  }
}

File: android/src/main/java/com/pitchdetector/PitchDetectorModule.kt
package com.pitchdetector

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.pitchdetector.processors.BaseProcessor

class PitchDetectorModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
  private val processor = BaseProcessor()
  private var recording = false

  override fun getName(): String {
    return NAME
  }

  @ReactMethod
  fun start(config: ReadableMap, promise: Promise) {
    if (!this.recording) {
      try {
        val emitter = reactContext.getJSModule(RCTDeviceEventEmitter::class.java)
        processor.start(config, emitter)
        this.recording = true
        promise.resolve(null)
      } catch (e: Exception) {
        this.recording = false
        promise.reject(e)
      }
    }
  }

  @ReactMethod
  fun stop(promise: Promise) {
    if (this.recording) {
      try {
        processor.stop()
        this.recording = false
        promise.resolve(null)
      } catch (e: Exception) {
        this.recording = false
        promise.reject(e)
      }
    }
  }

  @ReactMethod
  fun isRecording(promise: Promise) {
    try {
      promise.resolve(this.recording)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun addListener(eventName: String?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @ReactMethod
  fun removeListeners(count: Int?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  companion object {
    const val NAME = "PitchDetectorModule"
  }
}
